### PR TITLE
feat: Complete list ops

### LINKED ISSUE #174: add list data structure APIs to python SDK
Referencing https://www.notion.so/momentohq/SDK-API-Methods-Spec-cf12e4db2535405ca2acc3608814b083#9eb4f2ebdae44a19923e9142ebdd98ec

- [x] listConcatenateBack
- [x] listConcatenateFront
- [x] listPushBack
- [x] listPushFront
- [x] listFetch
- [x] listLength
- [x] listPopBack
- [x] listPopFront
- [x] listRemoveValue

### PR DESCRIPTION
**Pull request by schwern**
* list_remove_value
* list_pop_back/front
* list_push_back/front
* list_length
* test list ttl behavior
* lists accept iterable values

Closes #174

### CODE REVIEW

File: src/momento/internal/synchronous/_scs_data_client.py [hunk 1]
```diff
@@ -245,6 +270,173 @@ def list_fetch(self, cache_name: TCacheName, list_name: TListName) -> CacheListF
             self._log_request_error("list_fetch", e)
             return CacheListFetch.Error(convert_error(e))
 
+    def list_length(self, cache_name: TCacheName, list_name: TListName) -> CacheListLengthResponse:
+        try:
+            self._log_issuing_request("ListLength", {"list_name": str(list_name)})
+            _validate_cache_name(cache_name)
+            _validate_list_name(list_name)
+            request = _ListLengthRequest()
+            request.list_name = _as_bytes(list_name, "Unsupported type for list_name: ")
```
Comment (malandis) [inline src/momento/internal/synchronous/_scs_data_client.py:279]: Let's pull out the error message as it is repeated here

File: src/momento/internal/synchronous/_scs_data_client.py [hunk 2]
```diff
@@ -245,6 +270,173 @@ def list_fetch(self, cache_name: TCacheName, list_name: TListName) -> CacheListF
...
+            )
+            self._log_received_response("ListLength", {"list_name": str(request.list_name)})
+
+            type = response.WhichOneof("list")
+            if type == "missing":
+                return CacheListLength.Miss()
+            elif type == "found":
+                return CacheListLength.Hit(response.found.length)
+            else:
+                raise UnknownException("Unknown list field")
```
Comment (malandis) [inline src/momento/internal/synchronous/_scs_data_client.py:293]: Same with this string

File: src/momento/responses/list_data/list_length.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    @dataclass
+    class Hit(CacheListLengthResponse):
+        """Indicates the list exists and its length was fetched."""
+
+        length: int
+        """The number of values in the list.
+
+        Returns:
+            int
```
Comment (malandis) [inline src/momento/responses/list_data/list_length.py:32]: No need to document this as a return value; it's an attribute of the class. The first line should do

File: src/momento/responses/list_data/list_pop_back.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    """Groups all `CacheListPopBack` derived types under a common namespace."""
+
+    @dataclass
+    class Hit(CacheListPopBackResponse, ValueStringMixin):
+        """Indicates the request was successful."""
+
+        value_bytes: bytes
+        """The value returned from the cache for the specified key. Use the
+        `value_string` property to access the value as a string."""
+
```
Comment (malandis) [inline src/momento/responses/list_data/list_pop_back.py:31]: Let's make this comment more specific to list pop

File: src/momento/responses/list_data/list_pop_back.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    @dataclass
+    class Miss(CacheListPopBackResponse):
+        """Contains the results of a cache miss"""
+
```
Comment (malandis) [inline src/momento/responses/list_data/list_pop_back.py:35]: Same here re the comment

File: src/momento/simple_cache_client.py [hunk 6]
```diff
@@ -135,12 +148,6 @@ def create_cache(self, cache_name: str) -> CreateCacheResponse:
 
         Returns:
             CreateCacheResponse:
-
-            This response is resolved to one of:
-
-            - `CreateCache.Success`
-            - `CreateCache.AlreadyExists`
-            - `CreateCache.Error`
```
Comment (malandis) [inline src/momento/simple_cache_client.py:143]: Are these changes in the async code?

File: src/momento/typing.py [hunk 7]
```diff
@@ -13,8 +13,9 @@
 
 # List Types
 TListName = TCollectionName
-TListValuesBytes = List[bytes]
-TListValuesStr = List[str]
+TListValue = Union[str, bytes]
+TListValuesBytes = Sequence[bytes]
+TListValuesStr = Sequence[str]
```
Comment (malandis) [inline src/momento/typing.py:18]: Per the discussion, we should accept the broadest types when storing data, but when retrieving the list, give back a concrete list

**Review by pgautier404 — COMMENTED** [review]
LGTM when Michael is happy with it.

File: src/momento/simple_cache_client.py (hunk 6 above)
Comment (schwern) [reply src/momento/simple_cache_client.py:143]: No. When I deleted them earlier I forgot to make gen-sync before committing. I don't think we'll conflict, but I can put up a separate PR to fix it.

File: src/momento/internal/aio/_scs_data_client.py [hunk 8]
```diff
@@ -245,6 +270,173 @@ async def list_fetch(self, cache_name: TCacheName, list_name: TListName) -> Cach
...
+            request.list_name = _as_bytes(list_name, "Unsupported type for list_name: ")
+            response = await self._build_stub().ListPopBack(
+                request,
+                metadata=make_metadata(cache_name),
+                timeout=self._default_deadline_seconds,
+            )
+            self._log_received_response("ListPopBack", {"list_name": str(request.list_name)})
+
+            type = response.WhichOneof("list")
+            if type == "missing":
```
Comment (cprice404) [inline src/momento/internal/aio/_scs_data_client.py:313]: does the generated code not provide constants for these?

if not, should we consider creating our own constants for them?  The string equality checking make me a little sad

**Review by cprice404 — COMMENTED** [review]
one question about the `oneof` handling, otherwise lgtm when Michael L. is happy.

File: src/momento/internal/aio/_scs_data_client.py (hunk 8 above)
Comment (schwern) [reply src/momento/internal/aio/_scs_data_client.py:313]: Nope, that's the official way to do it.
https://developers.google.com/protocol-buffers/docs/reference/python-generated#oneof

I don't think we need constants, the wire protocol will remain backwards compatible.

File: src/momento/typing.py [hunk 9]
```diff
@@ -14,8 +14,11 @@
 # List Types
 TListName = TCollectionName
 TListValue = Union[str, bytes]
-TListValuesBytes = Sequence[bytes]
-TListValuesStr = Sequence[str]
-TListValues = Union[TListValuesBytes, TListValuesStr]
+TListValuesInputBytes = Iterator[bytes]
+TListValuesInputStr = Iterator[str]
+TListValuesInput = Union[TListValuesInputBytes, TListValuesInputStr]
```
Comment (malandis) [inline src/momento/typing.py:19]: Inputs should be `Iterable[TListValue]`, right? Ie support heterogeneous inputs

Comment (schwern) [reply src/momento/typing.py]: Just pushed a fix.

File: src/momento/typing.py [hunk 10]
```diff
@@ -22,8 +22,12 @@
 
 # List Types
 TListName = TCollectionName
-TListValuesBytes = List[bytes]
-TListValuesStr = List[str]
-TListValues = Union[TListValuesBytes, TListValuesStr]
+TListValue = Union[str, bytes]
+TListValuesInputBytes = Iterable[bytes]
+TListValuesInputStr = Iterable[str]
+TListValuesInput = Union[TListValuesInputBytes, TListValuesInputStr]
```
Comment (malandis) [inline src/momento/typing.py:28]: List value scan be heterogeneous on the way in, ie `Iterable[TListValue]`

**Review by malandis — APPROVED** [review]

**Pull request merged**